File: market-rotation-tracker/src/services/rotationTracker.ts
import { DailyBar, SectorScore, RotationResult } from '../types';

// Compute simple rate of change over n periods
function roc(bars: DailyBar[], periods: number): number {
    if (bars.length < periods + 1) return 0;
    const latest = bars[bars.length - 1].close;
    const prev = bars[bars.length - 1 - periods].close;
    return prev > 0 ? ((latest - prev) / prev) * 100 : 0;
}

// Relative strength: sector ROC minus benchmark ROC (SPY)
function relativeStrength(sectorBars: DailyBar[], spyBars: DailyBar[], periods = 20): number {
    return roc(sectorBars, periods) - roc(spyBars, periods);
}

export const SECTOR_UNIVERSE: { symbol: string; label: string }[] = [
    { symbol: 'SPY',  label: 'S&P 500 (Benchmark)' },
    { symbol: 'XLK',  label: 'Technology' },
    { symbol: 'XLF',  label: 'Financials' },
    { symbol: 'XLV',  label: 'Health Care' },
    { symbol: 'XLE',  label: 'Energy' },
    { symbol: 'XLI',  label: 'Industrials' },
    { symbol: 'XLC',  label: 'Communication Services' },
    { symbol: 'XLY',  label: 'Consumer Discretionary' },
    { symbol: 'XLP',  label: 'Consumer Staples' },
    { symbol: 'XLB',  label: 'Materials' },
    { symbol: 'XLU',  label: 'Utilities' },
    { symbol: 'XLRE', label: 'Real Estate' },
    { symbol: 'QQQ',  label: 'Nasdaq 100 (Growth)' },
    { symbol: 'IWM',  label: 'Russell 2000 (Small Cap)' },
    { symbol: 'SMH',  label: 'Semiconductors' },
];

export function scoreRotation(
    barsBySymbol: Map<string, DailyBar[]>,
    spyBars: DailyBar[]
): RotationResult {
    const scores: SectorScore[] = [];

    for (const sector of SECTOR_UNIVERSE) {
        const bars = barsBySymbol.get(sector.symbol);
        if (!bars || bars.length < 25) continue;

        const rs = relativeStrength(bars, spyBars, 20);
        const mom10 = roc(bars, 10);
        // Composite: 60% relative strength + 40% short-term momentum
        const composite = rs * 0.6 + mom10 * 0.4;
        // Normalise to 0-100 range (clamp between -20 and +20 raw)
        const score = Math.max(0, Math.min(100, 50 + composite * 2.5));

        scores.push({
            symbol: sector.symbol,
            label: sector.label,
            relativeStrength: parseFloat(rs.toFixed(2)),
            momentum: parseFloat(mom10.toFixed(2)),
            score: parseFloat(score.toFixed(1)),
            trend: rs > 1 ? 'strengthening' : rs < -1 ? 'weakening' : 'neutral',
        });
    }

    scores.sort((a, b) => b.score - a.score);

    const leaders = scores.slice(0, 3);
    const laggards = scores.slice(-3).reverse();
    const topMove = leaders[0]
        ? `${leaders[0].label} (${leaders[0].symbol}) leading with RS +${leaders[0].relativeStrength}% vs SPY`
        : 'No clear rotation leader';

    return {
        leaders,
        laggards,
        allScores: scores,
        topRotationMove: topMove,
        rankedAt: new Date().toISOString(),
    };
}

File: market-rotation-tracker-kotlin/src/main/kotlin/com/rotationtracker/services/RotationTracker.kt
package com.rotationtracker.services

import com.rotationtracker.models.DailyBar
import com.rotationtracker.models.RotationResult
import com.rotationtracker.models.SectorScore
import java.time.Instant

val SECTOR_UNIVERSE: List<Pair<String, String>> = listOf(
    "SPY"  to "S&P 500 (Benchmark)",
    "XLK"  to "Technology",
    "XLF"  to "Financials",
    "XLV"  to "Health Care",
    "XLE"  to "Energy",
    "XLI"  to "Industrials",
    "XLC"  to "Communication Services",
    "XLY"  to "Consumer Discretionary",
    "XLP"  to "Consumer Staples",
    "XLB"  to "Materials",
    "XLU"  to "Utilities",
    "XLRE" to "Real Estate",
    "QQQ"  to "Nasdaq 100 (Growth)",
    "IWM"  to "Russell 2000 (Small Cap)",
    "SMH"  to "Semiconductors",
)

/** Rate of change over n periods (%). */
private fun roc(bars: List<DailyBar>, periods: Int): Double {
    if (bars.size < periods + 1) return 0.0
    val latest = bars.last().close
    val prev = bars[bars.size - 1 - periods].close
    return if (prev > 0) ((latest - prev) / prev) * 100 else 0.0
}

/** Sector ROC minus SPY ROC over the lookback period. */
private fun relativeStrength(sectorBars: List<DailyBar>, spyBars: List<DailyBar>, periods: Int = 20): Double {
    return roc(sectorBars, periods) - roc(spyBars, periods)
}

fun scoreRotation(
    barsBySymbol: Map<String, List<DailyBar>>,
    spyBars: List<DailyBar>,
): RotationResult {
    val scores = mutableListOf<SectorScore>()

    for ((symbol, label) in SECTOR_UNIVERSE) {
        val bars = barsBySymbol[symbol] ?: continue
        if (bars.size < 25) continue

        val rs = relativeStrength(bars, spyBars, 20)
        val mom10 = roc(bars, 10)
        // 60% relative strength + 40% short-term momentum, normalised to 0–100
        val composite = rs * 0.6 + mom10 * 0.4
        val score = composite * 2.5 + 50.0

        scores.add(
            SectorScore(
                symbol          = symbol,
                label           = label,
                relativeStrength = "%.2f".format(rs).toDouble(),
                momentum        = "%.2f".format(mom10).toDouble(),
                score           = "%.1f".format(score.coerceIn(0.0, 100.0)).toDouble(),
                trend           = when {
                    rs > 1  -> "strengthening"
                    rs < -1 -> "weakening"
                    else    -> "neutral"
                },
            )
        )
    }

    scores.sortByDescending { it.score }

    val leaders  = scores.take(3)
    val laggards = scores.takeLast(3).reversed()
    val topMove  = leaders.firstOrNull()?.let {
        "${it.label} (${it.symbol}) leading with RS +${it.relativeStrength}% vs SPY"
    } ?: "No clear rotation leader"

    return RotationResult(
        leaders          = leaders,
        laggards         = laggards,
        allScores        = scores,
        topRotationMove  = topMove,
        rankedAt         = Instant.now().toString(),
    )
}
